datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Grouped natural id partitioner class that works with Hugging Face Datasets."""
+from typing import Any, Dict, List, Literal
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class GroupedNaturalIdPartitioner(Partitioner):
+    """Partition dataset by creating groups of natural ids.
+
+    Conceptually, you can think of this partitioner as a way of creating an organization
+    of x users instead of each user represetning a separate partition. You can change
+    the nature of the problem from cross-device to cross-silo (cross organization).
+
+    Parameters
+    ----------
+    partition_by: str
+        The name of the column that contains the unique values of partitions.
+    group_size: int
+        The number of unique ids that will be placed in a single group.
+    mode: Literal["allow-smaller", "allow-bigger", "drop-reminder", ""strict"]
+        The mode that will be used to handle the remainder of the unique ids.
+        - "allow-smaller": The last group can be smaller than the group_size.
+        - "allow-bigger": The first group can be bigger than the group_size.
+        - "drop-reminder": The last group will be dropped if it is smaller than the
+        group_size.
+        - "strict": Raises a ValueError if the remainder is not zero. In this mode you
+        expect each group to have the same size.
+    sort_unique_ids: bool
+        If True, the unique natural ids will be sorted before creating the groups.
+
+    Examples
+    --------
+    "sentiment140" (aka Twitter) dataset
+
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import GroupedNaturalIdPartitioner
+    >>>
+    >>> partitioner = GroupedNaturalIdPartitioner(partition_by="user", group_size=2)
+    >>> fds = FederatedDataset(dataset="sentiment140",
+    >>>                        partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    """
+
+    def __init__(
+        self,
+        partition_by: str,
+        group_size: int,
+        mode: Literal[
+            "allow-smaller", "allow-bigger", "drop-reminder", "strict"
+        ] = "allow-bigger",
```

Comment:
That's a good question. I went for "allow-bigger" because it's consistent with the default behavior of np.array_split. But that's not a reasonable or thought-through argument; I'm open to changing it. Wdyt?

Reply:
yeah. I think this is reasonable. happy to keep it as the default (in the end this shouldn't make a big diff generally speaking)